apollo-federation/src/sources/connect/models/validation.rs

```diff
@@ -132,114 +144,166 @@ fn validate_source(directive: &Component<Directive>) -> Result<SourceDirective,
-            Ok(Self::Valid(str_value.to_string()))
+            Self::Valid {
+                name: str_value.to_string(),
+                directive_name,
+            }
         } else {
-            Ok(Self::Invalid(str_value.to_string()))
+            Self::Invalid {
+                name: str_value.to_string(),
+                directive_name,
+            }
+        }
+    }
+
+    fn into_value_or_error(self) -> Result<String, ValidationError> {
+        match self {
+            Self::Valid { name, .. } => Ok(name),
+            Self::Invalid {
+                name,
+                directive_name,
+            } => Err(ValidationError::InvalidSourceName {
+                source_name: name,
+                directive_name,
+            }),
+            Self::Empty { directive_name } => {
+                Err(ValidationError::EmptySourceName { directive_name })
+            }
+            Self::Missing { directive_name } => Err(ValidationError::GraphQLError(format!(
+                "missing name argument to @{directive_name}"
+            ))),
         }
     }
 }
 
 impl Display for SourceName {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
         match self {
-            Self::Valid(name) | Self::Invalid(name) => write!(f, "{}", name),
+            Self::Valid {
+                name,
+                directive_name,
+            }
+            | Self::Invalid {
+                name,
+                directive_name,
+            } => write!(f, "@{directive_name} \"{name}\""),
+            Self::Empty { directive_name } | Self::Missing { directive_name } => {
+                write!(f, "unnamed @{directive_name}")
+            }
         }
     }
 }
 
 #[derive(Debug, thiserror::Error)]
 pub enum ValidationError {
     #[error("invalid GraphQL: {0}")]
-    GraphQLError(&'static str),
-    #[error("baseURL argument for @source \"{source_name}\" was not a valid URL: {inner}")]
+    GraphQLError(String),
+    #[error("baseURL argument for {source_name} was not a valid URL: {inner}")]
     SourceUrl {
         inner: url::ParseError,
         source_name: SourceName,
     },
-    #[error("baseURL argument for @source \"{source_name}\" must be http or https, got {scheme}")]
+    #[error("baseURL argument for {source_name} must be http or https, got {scheme}")]
     SourceScheme {
         scheme: String,
         source_name: SourceName,
     },
-    #[error("name argument for a @source directive must be a string")]
-    SourceNameType,
-    #[error("name argument to @source can't be empty")]
-    EmptySourceName,
+    #[error("name argument to @{directive_name} can't be empty")]
+    EmptySourceName { directive_name: DirectiveName },
     #[error(
-        "invalid characters in source name {0}, only alphanumeric and underscores are allowed"
+        "invalid characters in @{directive_name} name {source_name}, only alphanumeric and underscores are allowed"
```

Comment:
good thinking about renamed directives in error messages! 

what do you think about using the "schema coordinate" in the error message? i looked around the federation code base and saw messages like `Invalid default value (got: "foo") provided for argument Query.f(a:) of type Int.`

So this could be:
```suggestion
        "Invalid characters (got: "j$on") provided for argument @source(name:), only letters, numbers, and underscores are allowed"
```

Reply:
I do like that format here. As for more general coordinates, I'm going to need a general-purpose coordinate system _anyway_ to communicate the location of `@connect` errors back up to TypeScript. So, I'll definitely add a `Display` impl for that and include it in error messages.